core/coreapi/pin.go

```diff
@@ -191,45 +192,96 @@ func (p *pinInfo) Type() string {
 }
 
 func (api *PinAPI) pinLsAll(typeStr string, ctx context.Context) ([]coreiface.Pin, error) {
+	pinCh, errCh := PinLsAll(ctx, typeStr, api.pinning, api.dag)
+
+	var pins []coreiface.Pin
+loop:
+	for {
+		select {
+		case p, ok := <-pinCh:
+			if !ok {
+				break loop
+			}
+			pins = append(pins, p)
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case err := <-errCh:
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return pins, nil
+}
 
-	keys := make(map[cid.Cid]*pinInfo)
+// PinLsAll is an internal function for returning a list of pins
+func PinLsAll(ctx context.Context, typeStr string, pin pin.Pinner, dag ipld.DAGService) (chan coreiface.Pin, chan error) {
+	ch := make(chan coreiface.Pin, 32)
+	errCh := make(chan error, 1)
 
-	AddToResultKeys := func(keyList []cid.Cid, typeStr string) {
+	keys := cid.NewSet()
+	AddToResultKeys := func(keyList []cid.Cid, typeStr string) error {
 		for _, c := range keyList {
-			keys[c] = &pinInfo{
-				pinType: typeStr,
-				path:    path.IpldPath(c),
+			if keys.Visit(c) {
+				select {
+				case ch <- &pinInfo{
+					pinType: typeStr,
+					path:    path.IpldPath(c),
+				}:
+				case <-ctx.Done():
+					return ctx.Err()
+				}
 			}
 		}
+		return nil
 	}
 
-	if typeStr == "direct" || typeStr == "all" {
-		AddToResultKeys(api.pinning.DirectKeys(), "direct")
-	}
-	if typeStr == "indirect" || typeStr == "all" {
-		set := cid.NewSet()
-		for _, k := range api.pinning.RecursiveKeys() {
-			err := merkledag.Walk(
-				ctx, merkledag.GetLinksWithDAG(api.dag), k,
-				set.Visit,
-				merkledag.SkipRoot(), merkledag.Concurrent(),
-			)
-			if err != nil {
-				return nil, err
+	go func() {
+		defer close(ch)
```

Comment:
Given the way this function is currently called, this can hide errors: we can observe this channel closing before we read the error off the error channel.

It's generally safer to just use a "result" channel. That makes it easier to call this function as the `for x := range myCh` syntax can be used.

Reply:
Alternatively, we can close both the error and normal channels and, instead of listening on the error channel, run `err := <-errCh` after finishing the _normal_ loop.

Reply:
I think I did what you asked here, lmk.